VaccineCard/VaccineCard/Services/JWKSStorage.swift

```diff
@@ -0,0 +1,173 @@
+            print("name should be \n\(issuer.rawValue.filePathSafeName())")
+            #endif
+            return
+        }
+        do {
+            // Get data at path
+            let data = try Data(contentsOf: bundledFilePath)
+            // Decode
+            let keys = try JSONDecoder().decode(PublicKeys.self, from: data)
+            _ = store(keys: keys, for: issuer.rawValue)
+        } catch {
+            print(error.localizedDescription)
+            return
+        }
+    }
+    
+    
+    /// Fetch all stored jwks.json files for trusted issuers
+    /// - Returns: 2d array of issuer : jwks.json
+    public func fetchAll() -> [JWKSIssuers: PublicKeys]? {
+        let documentsDirectory = documentDirectory().appendingPathComponent("jwks")
+        guard directoryExists(path: documentsDirectory) else {return nil}
+        do {
+            let subDirectories = try FileManager.default.subpathsOfDirectory(atPath: documentsDirectory.path)
+            #if DEBUG
+            print(subDirectories)
+            #endif
+            var results: [JWKSIssuers: PublicKeys] = [JWKSIssuers: PublicKeys]()
+            for subDirectory in subDirectories where JWKSIssuers.allValues.contains(where: {$0.rawValue.filePathSafeName() == subDirectory}) {
+                let jwksPath = getDirectory(for: subDirectory).appendingPathComponent("jwks.json")
+                do {
+                    let data = try Data(contentsOf: jwksPath)
+                    // Decode
+                    let keys = try JSONDecoder().decode(PublicKeys.self, from: data)
+                    if let issuerEnum = JWKSIssuers.init(rawValue: subDirectory.filePathSafeNameToURL()) {
+                        results[issuerEnum] = keys
+                    }
+                } catch {
+                    print(error.localizedDescription)
+                    continue
+                }
+            }
+            return results
+        } catch {
+            print(error.localizedDescription)
+            return nil
+        }
+    }
+    
+    // MARK: Private fuctions
+    // These are wrapper functions for swift's FileManager
+    private func createDirectoryIfDoesntExist(path: URL) {
+        guard !directoryExists(path: path) else {return}
+        do {
+            try FileManager.default.createDirectory(atPath: path.path, withIntermediateDirectories: true, attributes: nil)
+        } catch let error as NSError {
+            print("Unable to create directory \(error.debugDescription)")
+        }
+    }
+    
+    private func getDirectory(for issuer: String) -> URL {
+        let documentsDirectory = documentDirectory().appendingPathComponent("jwks")
+        let dirPath = documentsDirectory.appendingPathComponent(issuer.filePathSafeName())
+        return dirPath
+    }
+    
+    private func documentDirectory() -> URL {
+        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
+        let documentsDirectory = paths[0]
```

Comment:
Only think I could think of would be to safely check that paths is not empty, to avoid an index out of range crash here - but I don't know how a user would be able to delete their documents directory, so would likely be an unnecessary check now that I think about it.